File: Hotel_Management_System/hotel_management_system.js
class Hotel {
    constructor(id, name, location, roomList = []) {
        this.id = id
        this.name = name
        this.location = location
        this.roomList = roomList
    }
}

class Location {
    constructor(pin, street, area, city, country){
        this.pin = pin
        this.street = street
        this.area = area
        this.city = city
        this.country = country
    }
}

class Room {
    constructor(roomId, roomType, roomStatus, price, roomKeys, houseKeepingLogs) {
        this.roomId = roomId
        this.roomType = roomType
        this.roomStatus = roomStatus
        this.price = price
        this.roomKeys = roomKeys
        this.houseKeepingLogs = houseKeepingLogs
    }
}

const RoomType = Object.freeze({
    standard: "standard",
    deluxe: "deluxe",
    family: "family"
})

const RoomStatus = Object.freeze({
    available: "available",
    booked: "booked",
    reserved: "reserved"
})

class RoomKey {
    constructor(keyId, barcode, issuedAt, isActive, isMaster) {
        this.keyId = keyId,
        this.barcode = barcode,
        this.issuedAt = issuedAt
        this.isActive = isActive,
        this.isMaster = isMaster
    }

    assignRoom() {

    }
}

class HouseKeepingLog {
    constructor(desc, startDate, duration, houseKeeper) {
        this.desc = desc
        this.startDate = startDate
        this.duration = duration
        this.houseKeeper = houseKeeper
    }

    updateLogs() {

    }
}

class Person {
    constructor(id, name, account){
        this.id = id
        this.name = name
        this.account = account
    }
}

class HouseKeeper extends Person {
    getRoomserviced(date) {

    }
}

class Account {
    constructor(accountId, username, password, accountStatus){
        this.accountId = accountId
        this.username = username
        this.password = password
        this.accountStatus = accountStatus
    }
}

const AccountStatus = Object.freeze({
    active: "active",
    blocked: "blocked",
    closed: "closed"
})

class Guest extends Person {
    constructor(id, name, account, search, booking) {
        super(id, name, account)
        this.search = search
        this.booking = booking
    }

    getAllRooms(room) {

    }
}

class Search {
    searchRoom(roomList, roomType, startDate, duration){
        return roomList.filter(room => {
            return room.roomType === roomType && 
                   room.roomStatus === RoomStatus.available
        });
    }
}

class Booking {
    bookRoom(guestInfo, room, startDate, duration) {
        if(room.roomStatus != RoomStatus.available) {
            console.log(`Room ${room.roomId} is not available.`);
            return null;
        }

        const bookingId = "B" + Math.floor(Math.random() * 10000);
        const booking = new RoomBooking(bookingId, startDate, duration, guestInfo, room, 0);
        room.roomStatus = RoomStatus.booked

        return booking
    }
    cancelRoom(room) {
        room.roomStatus = RoomStatus.available;
        console.log(`Room ${room.roomId} is now available again.`);
    }
}

class Admin extends Person {
    addRoom(roomDetails) {}
    removeRoom(roomDetails) {}
    editRoom(roomDetails) {}
}

class Receptionist extends Person {
    constructor(id, name, account, search, booking){
        super(id, name, account)
        this.search = search
        this.booking = booking
    }

    checkIn(guestInfo, bookingDetails) {}
    checkOut(guestInfo, bookingDetails) {}
}

class RoomBooking {
    constructor(bookingId, startDate, duration, guestInfo, roomInfo, totalCharge) {
        this.bookingId = bookingId
        this.startDate = startDate
        this.duration = duration
        this.guestInfo = guestInfo
        this.roomInfo = roomInfo
        this.totalCharge = totalCharge
    }

    printSummary() {
        console.log(`Booking ID: ${this.bookingId}`);
        console.log(`Guest: ${this.guestInfo.name}`);
        console.log(`Room: ${this.roomInfo.roomId}`);
        console.log(`Duration: ${this.duration} nights`);
        console.log(`Total Charge: ₹${this.totalCharge}`);
    }
}

class BaseRoomCharge {
    getCost() {

    }
}

class RoomCharge extends BaseRoomCharge {
    constructor(cost) {
        super()
        this.cost = cost
    }

    getCost() {
        return this.cost
    }

    setCost(cost) {
        this.cost = cost
    }
}

class RoomServiceCharge extends BaseRoomCharge {
    constructor(cost, baseroomcharge) {
        super();
        this.cost = cost;
        this.baseroomcharge = baseroomcharge;
    }

    getCost() {
        return this.baseroomcharge.getCost() + this.cost;
    }
}

class InRoomPurchase extends BaseRoomCharge {
    constructor(cost, baseroomcharge) {
        super();
        this.cost = cost;
        this.baseroomcharge = baseroomcharge;
    }

    getCost() {
        return this.baseroomcharge.getCost() + this.cost;
    }
}

const hotelLocation = new Location("560001", "MG Road", "Central", "Bangalore", "India");
const hotel = new Hotel(1, "The Grand Stay", hotelLocation);

const key1 = new RoomKey(1, "BAR123", new Date(), true, false);
const housekeepingLog = new HouseKeepingLog("Cleaned", new Date(), 30, null);

const room = new Room(
    "R101",
    RoomType.deluxe,
    RoomStatus.available,
    3000,
    [key1],
    [housekeepingLog]
);

hotel.roomList.push(room);

const guestAccount = new Account(1, "john_doe", "password123", AccountStatus.active);
const guest = new Guest(1, "John Doe", guestAccount, new Search(), new Booking());

const availableRooms = guest.search.searchRoom(hotel.roomList, RoomType.deluxe, new Date(), 3);

if (availableRooms.length === 0) {
    console.log("No rooms available.");
} else {
    const roomToBook = availableRooms[0];
    
    const booking = guest.booking.bookRoom(guest, roomToBook, new Date(), 3);

    if (booking) {
        const baseCharge = new RoomCharge(roomToBook.price * booking.duration);
        const withRoomService = new RoomServiceCharge(1000, baseCharge);
        const withInRoomPurchase = new InRoomPurchase(500, withRoomService);

        booking.totalCharge = withInRoomPurchase.getCost();

        console.log(`Hotel: ${hotel.name}`);
        console.log(`Guest: ${guest.name}`);
        console.log(`Room Booked: ${roomToBook.roomId}`);
        console.log(`Total Cost: ₹${booking.totalCharge}`);
        console.log(`Room Status: ${roomToBook.roomStatus}`);
    }
}

File: Hotel_Management_System/hotel_management_system.py
from dataclasses import dataclass, field
from typing import List, Optional
from enum import Enum
from datetime import date

@dataclass
class Location:
    pin: int
    street: str
    area: str
    city: str
    country: str

class RoomType(Enum):
    STANDARD = "STANDARD"
    DELUXE = "DELUXE"
    FAMILY = "FAMILY"

class RoomStatus(Enum):
    BOOKED = "BOOKED"
    AVAILABLE = "AVAILABLE"
    RESERVED = "RESERVED"

@dataclass
class RoomKey:
    keyId: int
    isActive: bool
    isMaster: bool

    def assignRoom(self, room: 'Room'):
        pass

class AccountStatus(Enum):
    ACTIVE = "ACTIVE"
    BLOCKED = "BLOCKED"
    CLOSED = "CLOSED"

@dataclass
class Account:
    id: int
    username: str
    password: str
    accountStatus: AccountStatus

@dataclass
class Person:
    id: int
    name: str
    account: Account

@dataclass
class HouseKeeper(Person):
    def getRoomServiced(self, date: date):
        return []


@dataclass
class HouseKeepingLog:
    description: str
    startDate: date
    duration: int
    houseKeeper: HouseKeeper

    def updateLogs(self, room: 'Room'):
        room.houseKeepingLogs.append(self)

@dataclass
class Room:
    roomId: int
    roomType: RoomType
    roomStatus: RoomStatus
    price: float
    roomKeys: List[RoomKey]
    houseKeepingLogs: List[HouseKeepingLog]


@dataclass
class Search:
    def searchRooms(self, roomType: RoomType, startDate: date, duration: int) -> List[Room]:
        return []

@dataclass
class Book:
    def bookRoom(self, guestInfo: 'Guest'):
        pass
    
    def cancelRoom(self, geustInfo: 'Guest'):
        pass

@dataclass
class Guest(Person):
    search: Search
    book: Book

    def getAllRooms(self, room: 'Room'):
        pass

@dataclass
class Admin(Person):
    def addRoom(self, roomDetails: 'Room'):
        pass

    def removeRoom(self, roomDetails: 'Room'):
        pass

    def editRoom(self, roomDetails: 'Room'):
        pass

class BookingStatus(Enum):
    BOOKED = "BOOKED"
    RESERVED = "RESERVED"
    CANCELED = "CANCELED"

class BaseRoomCharge:
    def getCost(self) -> float:
        raise NotImplementedError

@dataclass
class RoomCharge(BaseRoomCharge):
    cost: float
    def getCost(self) -> float:
        return self.cost
    
    def setCost(self, cost: float):
        self.cost = cost

@dataclass
class RoomServiceCharge(BaseRoomCharge):
    cost: float
    base_room_charge: BaseRoomCharge

    def getCost(self) -> float:
        self.base_room_charge.setCost(self.base_room_charge.getCost() + self.cost)
        return self.base_room_charge.getCost()

@dataclass
class InRoomPurchase(BaseRoomCharge):
    cost: float
    base_room_charge: BaseRoomCharge

    def getCost(self) -> float:
        self.base_room_charge.setCost(self.base_room_charge.getCost() + self.cost)
        return self.base_room_charge.getCost()

@dataclass
class RoomBooking:
    bookingId: int
    startDate: date
    duration: int
    bookingStatus: BookingStatus
    guestInfo: List['Guest']
    roomInfo: List[Room]
    totalCharge: BaseRoomCharge

@dataclass
class Receptionist(Person):
    search: Search
    book: Book

    def checkIn(self, guestInfo: 'Guest', bookingInfo: RoomBooking):
        pass

    def checkOut(self, guestInfo: 'Guest', bookingInfo: RoomBooking):
        pass

@dataclass
class Hotel:
    id: int
    name: str
    location: Location
    roomList: List[Room] = field(default_factory=list)
